File: backend/app/routers/scene.py
"""
场景相关路由

✅ 已实现：
- GET /api/scene/list 获取场景列表
- GET /api/scene/{id} 获取单个场景详情

参考实现见 frontend/src/api/scene.js 的注释
"""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.schemas import SceneInfo
from app.models import Scene, User
from app.database import get_db
from app.routers.auth import get_current_user

router = APIRouter(prefix="/api/scene", tags=["场景"])


def _scene_info(scene: Scene) -> SceneInfo:
    info = SceneInfo.model_validate(scene)
    if "独立看病" in (scene.title or ""):
        return info.model_copy(update={
            "description": "以江苏省人民医院为背景，体验第一次独立看病中的关键决策与沟通",
            "role": "医院流程角色",
        })
    return info


@router.get("/list", response_model=List[SceneInfo])
async def get_scene_list(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取场景列表

    逻辑：
    1. 查询 is_active=1 的所有场景
    2. 按难度升序排列
    3. 返回列表
    """
    scenes = (
        db.query(Scene)
        .filter(Scene.is_active == 1)
        .order_by(Scene.difficulty.asc())
        .all()
    )
    return [_scene_info(scene) for scene in scenes]


@router.get("/{scene_id}")
async def get_scene_detail(
    scene_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取单个场景详情

    逻辑：
    1. 按 id 查询场景 → 404
    2. 返回完整场景信息（含 config）
    """
    scene = db.query(Scene).filter(Scene.id == scene_id).first()
    if not scene:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="场景不存在")

    payload = {
        "id": scene.id,
        "title": scene.title,
        "description": scene.description,
        "cover": scene.cover,
        "role": scene.role,
        "role_avatar": scene.role_avatar,
        "background": scene.background,
        "opening_message": scene.opening_message,
        "config": scene.config,
        "difficulty": scene.difficulty,
    }
    if "独立看病" in (scene.title or ""):
        payload.update({
            "description": "以江苏省人民医院为背景，体验第一次独立看病中的关键决策与沟通",
            "role": "医院流程角色",
        })
    return payload

File: frontend/src/api/scene.js
/**
 * =====================================================
 * 场景相关 API
 * 后端需实现：Python FastAPI
 * 路径前缀：/api/scene
 *
 * 已接入：场景列表与详情加载
 * =====================================================
 */

import request from './request'

/**
 * @api {GET} /api/scene/list 获取场景列表
 * @apiName GetSceneList
 * @apiGroup Scene
 *
 * @apiHeader {String} Authorization Bearer token
 *
 * @apiSuccess {Array} scenes 场景列表
 * @apiSuccess {Number} scenes.id 场景 ID
 * @apiSuccess {String} scenes.title 场景标题
 * @apiSuccess {String} scenes.description 场景描述
 * @apiSuccess {String} scenes.cover 场景封面图 URL
 * @apiSuccess {String} scenes.role 角色信息
 * @apiSuccess {Number} scenes.difficulty 难度等级 1-5
 *
 * @apiExample 前端调用：
 * const scenes = await getSceneList()
 * // scenes = [
 * //   { id: 1, title: '异物梗阻急救', description: '...', cover: '...', role: '成人患者', difficulty: 3 },
 * //   { id: 2, title: '第一次独立看病', description: '...', cover: '...', role: '青少年患者', difficulty: 2 }
 * // ]
 */
export function getSceneList() {
  return request({
    url: '/scene/list',
    method: 'GET'
  })
}

/**
 * @api {GET} /api/scene/:id 获取单个场景详情
 * @apiName GetSceneDetail
 * @apiGroup Scene
 *
 * @apiHeader {String} Authorization Bearer token
 *
 * @apiParam {Number} id 场景 ID
 *
 * @apiSuccess {Object} scene 场景详情
 * @apiSuccess {Number} scene.id 场景 ID
 * @apiSuccess {String} scene.title 场景标题
 * @apiSuccess {String} scene.description 场景描述
 * @apiSuccess {String} scene.background 场景背景
 * @apiSuccess {String} scene.role 角色信息
 * @apiSuccess {String} scene.role_avatar 角色头像 URL
 * @apiSuccess {String} scene.opening_message 角色开场白
 * @apiSuccess {Object} scene.config 场景配置
 */
export function getSceneDetail(id) {
  return request({
    url: `/scene/${id}`,
    method: 'GET'
  })
}
